File: rd-net/RdFramework/Base/ISingleContextHandler.cs
using JetBrains.Serialization;

namespace JetBrains.Rd.Base
{
  internal interface ISingleContextHandler
  {
    void WriteValue(SerializationCtx context, UnsafeWriter writer);

    void RegisterValueInValueSet();

    object ReadValueBoxed(SerializationCtx context, UnsafeReader reader);
    RdContextBase ContextBase { get; }
  }
  
  internal interface ISingleContextHandler<T> : ISingleContextHandler
  {
    RdContext<T> Context { get; }

    T? ReadValue(SerializationCtx context, UnsafeReader reader);
  }
}

File: rd-kt/rd-framework/src/commonMain/kotlin/com/jetbrains/rd/framework/base/ISingleContextHandler.kt
package com.jetbrains.rd.framework.base

import com.jetbrains.rd.framework.AbstractBuffer
import com.jetbrains.rd.framework.RdContext
import com.jetbrains.rd.framework.SerializationCtx

internal interface ISingleContextHandler<T: Any> {
    val context: RdContext<T>

    @Suppress("UNCHECKED_CAST")
    fun writeValue(ctx: SerializationCtx, buffer: AbstractBuffer)

    fun readValue(ctx: SerializationCtx, buffer: AbstractBuffer): T?

    fun registerValueInValueSet()
}
